packages/aws-cdk/lib/api/deployments.ts

```diff
@@ -469,31 +430,28 @@ export class Deployments {
       };
     }
 
-    const {
-      stackSdk,
-      resolvedEnvironment,
-      cloudFormationRoleArn,
-      envResources,
-    } = await this.prepareSdkFor(options.stack, options.roleArn, Mode.ForWriting);
+    const env = await this.envs.accessStackForStackOperations(options.stack, Mode.ForWriting);
 
     // Do a verification of the bootstrap stack version
     await this.validateBootstrapStackVersion(
       options.stack.stackName,
       options.stack.requiresBootstrapStackVersion,
       options.stack.bootstrapStackVersionSsmParameter,
-      envResources);
+      env.resources);
+
+    const executionRoleArn = await env.replacePlaceholders(options.roleArn ?? options.stack.cloudFormationExecutionRoleArn);
```

Comment:
Danger to forget the call to `replacePlaceholders()`

Reply:
Fair point, but I genuinely don't know how to do it better. 

The only way I can think of would be to start using branded types.